obfsproxy: new package

jefferyto opened the pull request:
From the [Tor project page](https://www.torproject.org/projects/obfsproxy.html.en):

obfsproxy is a tool that attempts to circumvent censorship, by
transforming the Tor traffic between the client and the bridge. This
way, censors, who usually monitor traffic between the client and the
bridge, will see innocent-looking transformed traffic instead of the
actual Tor traffic.

This depends on:

- pyptlib (#2053)
- twisted (#2052)

Also, txsocksx (#2058) is necessary to use an outgoing SOCKS proxy,
and having either gmpy2 (#2067) or gmpy (#2051) installed will help
speed up calculations.

Signed-off-by: Jeffery To <jeffery.to@gmail.com>

vantonio commented:
@jefferyto I just finished building with this, but the resulting package seemed to malformed.

`
> tar xvf obfsproxy_0.2.13-1_ramips_24kec.tar
./debian-binary
./data.tar.gz
./control.tar.gz
> tar tvf data.tar.gz
drwxr-xr-x root/root         0 2016-02-27 23:03 ./
drwxr-xr-x root/root         0 2016-02-27 23:03 ./usr/
drwxr-xr-x root/root         0 2016-02-27 23:03 ./usr/bin/
-rwxr-xr-x root/root       314 2016-02-27 23:03 ./usr/bin/obfsproxy
drwxr-xr-x root/root         0 2016-02-27 23:03 ./etc/
drwxr-xr-x root/root         0 2016-02-27 23:03 ./etc/init.d/
-rwxr-xr-x root/root      3505 2016-02-27 23:03 ./etc/init.d/obfsproxy
drwxr-xr-x root/root         0 2016-02-27 23:03 ./etc/config/
-rw------- root/root      1500 2016-02-27 23:03 ./etc/config/obfsproxy`

Seems like the resulting package has only skeletons of the obfsproxy library, complaints `The 'obfsproxy==0.2.13' distribution was not found and is required by the application` when you run it. 

Am I doing something wrong or can you offer some help on this? Thanks.

jefferyto commented:
Just noticed that there was an obfsproxy package in the old packages feed, so technically this isn't a "new" package (although I didn't base any of this work on the older files).

jefferyto commented:
@vantonio Did you build the dependencies as well? There are a number of new packages (in preceding pull requests) that are required for obfsproxy to work (and possibly build).

yousong commented on net/obfsproxy/files/obfsproxy.conf:
```diff
@@ -0,0 +1,57 @@
+	option listen_port '8080'
+
+	# Set to log to a file instead of syslog
+	#option log_file '/var/log/obfsproxy.log'
```
Obfsproxy writes log to stdout by default.  Does it support writing to syslog?

jefferyto replied:
It doesn't support writing to syslog by itself, but as procd is configured to route stdout to syslog, the comment is phrased in this way to convey the end result.

yousong replied:
Okay, then the comment line does make sense.

yousong commented on net/obfsproxy/files/obfsproxy.conf:
```diff
@@ -0,0 +1,57 @@
+	#option no_safe_logging 0
+
+	# Run as a different user
+	#option user 'obfsproxy'
```
How about defaulting to `nobody`?

jefferyto replied:
You mean use `nobody` as the default user, even when the `user` option is not specified, right? AFAIK the "natural" expectation for daemons (rightly or wrongly) is for them to run as `root` by default and to have an option to run them as another user, rather than the other way around, but I don't have a strong preference either way.

yousong replied:
Daemons should run with, or drop to retain only minimal privileges.  Well, it's true that many packages do not specifically enforce on this, it's just still the best practice.

If user 'obfsproxy' is hinted, then we'd better create such a user to make the default consistent.

jefferyto replied:
A "obfsproxy" user is hinted only because I prefer using different users for different daemons. It can be changed to "nobody" if it will make things easier.

yousong commented on net/obfsproxy/files/obfsproxy.init:
```diff
@@ -0,0 +1,153 @@
+
+	procd_set_param respawn
+	procd_set_param stdout 1
+	procd_set_param stderr 1
```
Logging `stderr` is only useful when something bad happens when obfsproxy tries to parsing the command line.  And logging to stdout should be disabled if `--log-file` is present.

jefferyto replied:
If `--log-file` is present, then obfsproxy should output to the log file rather than to `stdout`. So capturing `stdout` in this case should make no difference, unless there are other factors I'm not aware of?

yousong replied:
It's just one nitpicking to try to keep unnecessary parts trimmed out :)  Otherwise procd will need to unnecessarily keep track of 2 additional ustream bufs.

yousong commented on net/obfsproxy/files/obfsproxy.init:
```diff
@@ -0,0 +1,153 @@
+	config_get_bool enabled "$cfg" 'enabled' '0'
+	[ "$enabled" = 0 ] && return 1
+
+	config_get user "$cfg" 'user' 'root'
```
Keeping the "def" user same as the one in the sample config can be good.

yousong commented:
Hi, @jefferyto , finally we are at this package after merging all its dependencies :)

Personally I still do not like the idea of using such "verbose" uci config file for just args translation purposes [1].  It may help GUI users, but I also have doubts about that...  Surely it's hard (if not impossible) for one flavour of config file to please everyone, and it's also about making choices for future possible users [2,3] because once we make it into repo we need to maintain backward compatibility.  So I am going to hold the uci part for a while to let other people here have a chance to review.

[1] https://github.com/openwrt/packages/pull/2403#issuecomment-187038563
[2] https://github.com/openwrt/packages/blob/master/net/shadowsocks-libev/files/shadowsocks-libev.config
[3] https://github.com/yousong/waller/blob/master/shadowsocks-libev/files/shadowsocks-libev.config

jefferyto commented:
Thanks again for all your help @yousong!

I'd be happy to wait for more input. FWIW every package I've configured as a user has had this "verbose" format; I almost always have both the relevant uci wiki reference page and the original man page open at the same time.

The verbose uci format is definitely a leaky abstraction, but it seems to be the agreed-upon approach, at least for now. I think the "verbose vs concise" discussion is worth having for the project as a whole; I would just want this config file to follow the project's conventions.

yousong commented:
The problem with obfsproxy in this case is that it does not have a manpage at all, or did I just missed it?  Hah, it's not relevant and I digressed :)

kkrrll commented:
@jefferyto , thanks for all you do!  I finish the building with all dependencies and scp all to my router. after reboot, it reports errors like below, any help? 

Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]: Traceback (most recent call last):
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/bin/obfsproxy", line 9, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     load_entry_point('obfsproxy==0.2.13', 'console_scripts', 'obfsproxy')()
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/setuptools-7.0-py2.7.egg/pkg_resources.py", line 339, in load_entry_point
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/setuptools-7.0-py2.7.egg/pkg_resources.py", line 2470, in load_entry_point
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/setuptools-7.0-py2.7.egg/pkg_resources.py", line 2184, in load
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/obfsproxy/pyobfsproxy.py", line 12, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     import obfsproxy.network.launch_transport as launch_transport
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/obfsproxy/network/launch_transport.py", line 2, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     import obfsproxy.transports.transports as transports
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/obfsproxy/transports/transports.py", line 4, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     import obfsproxy.transports.obfs2 as obfs2
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/obfsproxy/transports/obfs2.py", line 13, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     import obfsproxy.common.aes as aes
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/obfsproxy/common/aes.py", line 6, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     from Crypto.Cipher import AES
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:   File "/usr/lib/python2.7/site-packages/Crypto/Cipher/AES.py", line 50, in <module>
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]:     from Crypto.Cipher import _AES
Sun Mar 27 12:27:05 2016 daemon.err obfsproxy[1352]: ImportError: File not found
Sun Mar 27 12:27:05 2016 daemon.info procd: Instance obfsproxy::instance1 s in a crash loop 6 crashes, 17 seconds since last crash _

yousong commented:
Hi, @kkrrll , all dependencies of obfsproxy should now already be merged into the master branch.  As for the exception you experienced above, please check if there exists such a file as `/usr/lib/python2.7/site-packages/Crypto/Cipher/_AES.so` on your router.  It should be built and installed as part of the `python-crypto` package.

kkrrll commented:
@yousong, thank you. I checked it, There are:

> root@OpenWrt:~# ls -l /usr/lib/python2.7/site-packages/Crypto/Cipher/
-rw-r--r--    1 root     root          4218 Mar 25 12:07 AES.py
-rw-r--r--    1 root     root          4917 Mar 25 12:07 ARC2.py
-rw-r--r--    1 root     root          4266 Mar 25 12:07 ARC4.py
-rw-r--r--    1 root     root          4460 Mar 25 12:07 Blowfish.py
-rw-r--r--    1 root     root          4498 Mar 25 12:07 CAST.py
-rw-r--r--    1 root     root          4403 Mar 25 12:07 DES.py
-rw-r--r--    1 root     root          5130 Mar 25 12:07 DES3.py
-rw-r--r--    1 root     root          9373 Mar 25 12:07 PKCS1_OAEP.py
-rw-r--r--    1 root     root          9103 Mar 25 12:07 PKCS1_v1_5.py
-rw-r--r--    1 root     root          2737 Mar 25 12:07 XOR.py
-rwxr-xr-x    1 root     root         33456 Mar 25 12:07 _AES.so
-rwxr-xr-x    1 root     root         17072 Mar 25 12:07 _ARC2.so
-rwxr-xr-x    1 root     root          8804 Mar 25 12:07 _ARC4.so
-rwxr-xr-x    1 root     root         33456 Mar 25 12:07 _Blowfish.so
-rwxr-xr-x    1 root     root         29360 Mar 25 12:07 _CAST.so
-rwxr-xr-x    1 root     root         53940 Mar 25 12:07 _DES.so
-rwxr-xr-x    1 root     root         53940 Mar 25 12:07 _DES3.so
-rwxr-xr-x    1 root     root          8808 Mar 25 12:07 _XOR.so
-rw-r--r--    1 root     root          3440 Mar 25 12:07 __init__.py
-rw-r--r--    1 root     root         12596 Mar 25 12:07 blockalgo.py

yousong commented:
No, `_AES.so` was not present, but I am also not sure what's that `AES.so` for.  Please check your build environment and `make clean` if you deem necessary ;)  I have built python-crypto here for mips-malta-be, `_AES.so` is there and obfsproxy worked fine.

kkrrll commented:
@yousong , there's _AES.so itself. I don't know why display lost  "_" after I Ctrl+x, the file do exist.

yousong commented:
@kkrrll This is more like a python-specific problem.  Is it possible that the `_AES.so` was not built for `ar71xx`?  I got the following lines in the output of command `strace python -c 'from Crypto.Cipher import AES' 2>/tmp/x`

```
stat64("/usr/lib/python2.7/site-packages/Crypto/Cipher/_AES", 0x7ff960f0) = -1 ENOENT (No such file or directory)
open("/usr/lib/python2.7/site-packages/Crypto/Cipher/_AES.so", O_RDONLY|O_LARGEFILE) = 4
fstat64(4, {st_mode=S_IFREG|0755, st_size=33456, ...}) = 0
open("/usr/lib/python2.7/site-packages/Crypto/Cipher/_AES.so", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 5
```

As for the missing underline character, that's because `_` is a markup character for _italicizing_ the content within.  You can markup block of code here with three continuous backquotes

kkrrll commented:
@yousong Thanks a lot for all your help! I found [this](http://stackoverflow.com/questions/26943136/python-aes-import-error-please) and [this](https://dev.openwrt.org/ticket/19893) may be related. All my packages are built under newer trunk and running in a chaos_calmer rom. May that break compatibility?

yousong commented on net/obfsproxy/files/obfsproxy.conf:
```diff
@@ -0,0 +1,57 @@
+	#option proxy_username ''
+	#option proxy_password ''
+	#option proxy_host ''
+	#option proxy_port ''
```
We can definitely simplify this part to a single option `proxy`

yousong commented:
Hi @jefferyto , I'd like to suggest a patch for using syslog directly.  By doing this we do not depend on stdio logging provided by procd, and the `logread` output will have proper format.

```
--- obfsproxy/pyobfsproxy.py.orig	2016-03-30 13:39:13.020170572 +0800
+++ obfsproxy/pyobfsproxy.py	2016-03-30 13:38:48.252162820 +0800
@@ -42,6 +42,7 @@ def set_up_cli_parsing():
 
     parser.add_argument('-v', '--version', action='version', version=__version__)
     parser.add_argument('--log-file', help='set logfile')
+    parser.add_argument('--syslog', action='store_true', help='use syslog')
     parser.add_argument('--log-min-severity',
                         choices=['error', 'warning', 'info', 'debug'],
                         help='set minimum logging severity (default: %(default)s)')
@@ -108,12 +109,14 @@ def do_external_mode(args):
 def consider_cli_args(args):
     """Check out parsed CLI arguments and take the appropriate actions."""
 
-    if args.log_file:
+    if args.no_log:
+        log.disable_logs()
+    elif args.log_file:
         log.set_log_file(args.log_file)
+    elif args.syslog:
+        log.use_syslog()
     if args.log_min_severity:
         log.set_log_severity(args.log_min_severity)
-    if args.no_log:
-        log.disable_logs()
     if args.no_safe_logging:
         log.set_no_safe_logging()
 
--- obfsproxy/common/log.py.orig	2016-03-30 13:30:14.896002146 +0800
+++ obfsproxy/common/log.py	2016-03-30 13:45:45.060293276 +0800
@@ -1,5 +1,6 @@
 """obfsproxy logging code"""
 import logging
+import logging.handlers
 import sys
 
 from twisted.python import log
@@ -33,6 +34,11 @@ class ObfsLogger(object):
         self.obfslogger.addHandler(self.default_handler)
         self.obfslogger.propagate = False
 
+    def use_syslog(self):
+        self.obfslogger.removeHandler(self.default_handler)
+        syslog_handler = logging.handlers.SysLogHandler(address='/dev/log')
+        self.obfslogger.addHandler(syslog_handler)
+
     def set_formatter(self, handler):
         """Given a log handler, plug our custom formatter to it."""
 

 ```

vantonio commented:
@jefferyto, any update on this? I really would like to see this get merged!

jefferyto commented:
Sorry - will try to find time for this very soon.

yousong commented:
@jefferyto updates?

jefferyto commented:
@yousong I've incorporated a version of your syslog patch, and changed the config file commented-out user to `nobody`. I kept stdout/stderr redirection for when `log_min_severity` is set to `debug` because it would be difficult to debug if you can't see any stdout/stderr output.

The default user in the init file is `root` because the process is suppose to run as root if the user doesn't specifically set a user. I prefer splitting the proxy option into multiple fields and having the init file handle the proxy option syntax; this leaves one less thing for the user to get wrong, is clearer for the user and/or if the file is under version control.

If there are more minor changes to be made, I would be happy to address them after this package has been merged.

yousong commented:
Thanks.
